macos/FluentUI/Button.swift

```diff
@@ -0,0 +1,241 @@
+			}
+		}
+	}
+	
+	override public var title: String {
+		willSet {
+			guard wantsLayer == true else {
+				preconditionFailure("wantsLayer must be set so that the title is rendered on the layer")
+			}
+		}
+		didSet {
+			self.attributedTitle = NSAttributedString(string: title, attributes: [.foregroundColor : textColor])
+		}
+	}
+	
+	override public func updateTrackingAreas() {
+		super.updateTrackingAreas()
+		
+		// Remove existing trackingArea
+		if let trackingArea = trackingArea {
+			removeTrackingArea(trackingArea)
+			self.trackingArea = nil
+		}
+		
+		// Create a new trackingArea
+		let opts: NSTrackingArea.Options = [.mouseEnteredAndExited, .activeAlways]
+		let trackingArea = NSTrackingArea(
+			rect: bounds,
+			options: opts,
+			owner: self,
+			userInfo: nil
+		)
+		addTrackingArea(trackingArea)
+		self.trackingArea = trackingArea
+	}
+	
+	open override func mouseEntered(with event: NSEvent) {
+		mouseEntered = true
+		needsDisplay = true
+	}
+
+	open override func mouseExited(with event: NSEvent) {
+		mouseEntered = false
+		needsDisplay = true
+	}
+	
+	open override func mouseDown(with event: NSEvent) {
+		mouseDown = true
+		needsDisplay = true
+	}
+	
+	open override func mouseUp(with event: NSEvent) {
+		mouseDown = false
+		needsDisplay = true
+	}
+	
+	open override func updateLayer() {
+		if let layer = layer {
+			layer.borderWidth = borderWidth
+			layer.cornerRadius = cornerRadius
+			layer.backgroundColor = layerBackgroundColor.cgColor
+			layer.borderColor = outlineColor.cgColor
+		}
+	}
+	
+	open override var intrinsicContentSize: NSSize {
+		var intrinsicContentSize = super.intrinsicContentSize;
+
+		intrinsicContentSize.width = intrinsicContentSize.width + (horizontalPadding * 2)
+		intrinsicContentSize.height += (verticalPadding * 2);
+
+		return intrinsicContentSize;
+	}
+	
+	override public var wantsUpdateLayer: Bool {
+		return true
+	}
+	
+	/// The primary color of the button, AKA, the fill color in the primaryFilled style, and the outline in the primaryOutline style
+	@objc public var primaryColor: NSColor = defaultPrimaryColor {
+		didSet {
+			if primaryColor != oldValue {
+				needsDisplay = true
+			}
+		}
+	}
+
+	/// The primary style of the button
+	@objc public var style: ButtonStyle = .primaryFilled {
+		didSet {
+			if style != oldValue {
+				needsDisplay = true
+			}
+		}
+	}
+	
+	private func disabledColor(for color: NSColor) -> NSColor {
+		if #available(macOS 10.14, *) {
+			return color.withSystemEffect(.disabled)
+		} else {
+			return color.withAlphaComponent(0.25)
+		}
+	}
+	
+	private var trackingArea: NSTrackingArea?
+	
+	private var mouseEntered = false
+	private var mouseDown = false
+	
+	private var fillColor: NSColor {
+		return style == ButtonStyle.primaryFilled ? primaryColor : .clear
+	}
+	
+	private var outlineColor: NSColor {
+		let baseOutlineColor = style == ButtonStyle.primaryOutline ? primaryColor.withAlphaComponent(0.4) : .clear
```

Comment:
nit: should some of these alpha values be constants?

Reply:
By having these colors all on the bottom of my swift file, I was effectively treating them as constants. Furthermore, this control will have to implement the color providing protocol @anandrajeswaran added, so all of this will change regardless. Given there's only 4, it should be easy enough for me to pull them out as real constants